drivers/fuel_gauge/bq27z746/emul_bq27z746.c

```diff
@@ -0,0 +1,352 @@
+	/** I2C address of emulator */
+	uint16_t addr;
+};
+
+static int emul_bq27z746_read_altmac(const struct emul *target, uint8_t *buf, size_t len)
+{
+	const uint8_t manufacturer_name[] = "Texas Instruments";
+	const uint8_t device_name[] = "BQ27Z746";
+	const uint8_t device_chemistry[] = "LION";
+	const struct bq27z746_emul_data *data = target->data;
+
+	if (len < BQ27Z746_MAC_COMPLETE_LEN) {
+		LOG_ERR("When reading the ALTMAC, one must read the full %u byte",
+			BQ27Z746_MAC_COMPLETE_LEN);
+		return -EIO;
+	}
+
+	memset(buf, 0, len);
+
+	/*
+	 * The data read from BQ27Z746_ALTMANUFACTURERACCESS is:
+	 * 0..1: The command (for verification)
+	 * 2..33: The data
+	 * 34: Checksum calculated as (uint8_t)(0xFF - (sum of all command and data bytes))
+	 * 35: Length including command, checksum and length (e.g. data length + 4)
+	 */
+
+	/* Put the command in the first two byte */
+	sys_put_le16(data->mac_cmd, buf);
+
+	/* Based on the command, put some data and the length into the buffer. */
+	/* In all of the operations, don't consider the zero-terminator. */
+	switch (data->mac_cmd) {
+	case BQ27Z746_MAC_CMD_MANUFACTURER_NAME:
+		memcpy(&buf[2], manufacturer_name, sizeof(manufacturer_name) - 1);
+		buf[35] = sizeof(manufacturer_name) - 1 + BQ27Z746_MAC_OVERHEAD_LEN;
+		break;
+	case BQ27Z746_MAC_CMD_DEVICE_NAME:
+		memcpy(&buf[2], device_name, sizeof(device_name) - 1);
+		buf[35] = sizeof(device_name) - 1 + BQ27Z746_MAC_OVERHEAD_LEN;
+		break;
+	case BQ27Z746_MAC_CMD_DEVICE_CHEM:
+		memcpy(&buf[2], device_chemistry, sizeof(device_chemistry) - 1);
+		buf[35] = sizeof(device_chemistry) - 1 + BQ27Z746_MAC_OVERHEAD_LEN;
+		break;
+	default:
+		LOG_ERR("ALTMAC command 0x%x is not supported", data->mac_cmd);
+		return -EIO;
+	}
+
+	/* Calculate the checksum */
+	uint8_t sum = 0; /* Intentionally 8 bit wide and overflowing */
+
+	for (int i = 0; i < BQ27Z746_MAC_COMPLETE_LEN - 2; i++) {
+		sum += buf[i];
+	}
+	buf[34] = 0xFF - sum;
+
+	return 0;
+}
+
+static int emul_bq27z746_write(const struct emul *target, uint8_t *buf, size_t len)
+{
+	struct bq27z746_emul_data *data = target->data;
+	const uint8_t reg = buf[0];
+
+	switch (reg) {
+	case BQ27Z746_ALTMANUFACTURERACCESS:
+		data->mac_cmd = sys_get_le16(&buf[1]);
+		return 0;
+	default:
+		LOG_ERR("Writing is only supported to ALTMAC currently");
+		return -EIO;
+	}
+}
+
+static int emul_bq27z746_reg_read(const struct emul *target, int reg, int *val)
+{
+	switch (reg) {
+	case BQ27Z746_MANUFACTURERACCESS:
+		*val = 1;
+		break;
+	case BQ27Z746_ATRATE:
+		*val = -1;
+		break;
+	case BQ27Z746_ATRATETIMETOEMPTY:
+		*val = 1;
+		break;
+	case BQ27Z746_TEMPERATURE:
+		*val = 1;
+		break;
+	case BQ27Z746_VOLTAGE:
+		*val = 1;
+		break;
+	case BQ27Z746_BATTERYSTATUS:
+		*val = 1;
+		break;
+	case BQ27Z746_CURRENT:
+		*val = -1;
+		break;
+	case BQ27Z746_REMAININGCAPACITY:
+		*val = 1;
+		break;
+	case BQ27Z746_FULLCHARGECAPACITY:
+		*val = 1;
+		break;
+	case BQ27Z746_AVERAGECURRENT:
+		*val = -1;
```

Comment:
Why did we change this from `1` to `-1`? I'd rather not use `-1` since it's bit representation is a palindrome.

Reply:
My mistake trying to write a test for https://github.com/zephyrproject-rtos/zephyr/pull/59382#issuecomment-1632692529.

Changing it to -2.